controllers/certificate/certificate_controller.go

```diff
@@ -0,0 +1,170 @@
+package certificate
+
+import (
+	"context"
+	"crypto/x509"
+	"fmt"
+	"strings"
+	"time"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
+	"github.com/kyma-project/api-gateway/controllers"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/cert"
+	ctrl "sigs.k8s.io/controller-runtime"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	maxAge       = time.Hour * 24 * 365 // issue certificate with 365 days validity
+	untilRenewal = time.Hour * 24 * 14  // renew certificate 14 days before expiration
+
+	certificateName = "tls.crt"
+	keyName         = "tls.key"
+
+	secretNamespace = "kyma-system"
+	secretName      = "api-gateway-webhook-certificate"
+	serviceName     = "api-gateway-webhook-service"
+
+	APIRuleCRDName = "apirules.gateway.kyma-project.io"
+)
+
+func NewCertificateReconciler(mgr manager.Manager, reconciliationInterval time.Duration) *CertificateReconciler {
+	return &CertificateReconciler{
+		Client:                 mgr.GetClient(),
+		Scheme:                 mgr.GetScheme(),
+		log:                    mgr.GetLogger().WithName("certificate-controller"),
+		reconciliationInterval: reconciliationInterval,
+	}
+}
+
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="apiextensions.k8s.io",resources=customresourcedefinitions,verbs=get;list;watch;create;update;patch;delete
+
+func (r *CertificateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.log.Info("Received reconciliation request", "name", req.Name)
+
+	certificateSecret := &corev1.Secret{}
+	err := r.Client.Get(ctx, req.NamespacedName, certificateSecret)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.log.Info("Reconciling Webhook Secret", "name", certificateSecret.Name)
+
+	err = verifySecret(certificateSecret)
+	if err == nil {
+		r.log.Info("Secret certificate is still valid and does not need to be updated")
+	} else {
+		r.log.Info("Secret certificate is invalid", "verificationError", err.Error())
+		certificate, err := createNewSecret(ctx, r.Client, certificateSecret)
+		if err != nil {
+			return ctrl.Result{RequeueAfter: 1 * time.Minute}, err
```

Comment:
Why do we set a RequeueAfter? If we don't set it, we can just use the default with an increasing back-off.

Reply:
Will change it to `return ctrl.Result{Requeue: true}, err` and rely on exponential back-off settings.

Reply:
It is an error scenario when certificate can not be generated / update and this is downtime for the conversion webhook, so it is important to be solved quickly.